lib/irb/color.rb

```diff
@@ -170,33 +165,37 @@ def without_circular_ref(obj, seen:, &block)
       end
 
       def scan(code, allow_last_error:)
-        pos = [1, 0]
-
         verbose, $VERBOSE = $VERBOSE, nil
         RubyLex.compile_with_errors_suppressed(code) do |inner_code, line_no|
           lexer = Ripper::Lexer.new(inner_code, '(ripper)', line_no)
+          byte_pos = 0
+          line_positions = [0]
+          inner_code.lines.each do |line|
+            line_positions << line_positions.last + line.bytesize
+          end
+
+          on_scan = proc do |elem|
+            str = elem.tok
+            start_pos = line_positions[elem.pos[0] - 1] + elem.pos[1]
+            end_pos = start_pos + str.bytesize
+            next if start_pos < byte_pos
+
+            yield(nil, inner_code.byteslice(byte_pos...start_pos), nil) if byte_pos < start_pos
+            yield(elem.event, str, elem.state)
+            byte_pos = end_pos
+          end
+
           if lexer.respond_to?(:scan) # Ruby 2.7+
             lexer.scan.each do |elem|
-              str = elem.tok
               next if allow_last_error and /meets end of file|unexpected end-of-input/ =~ elem.message
-              next if ([elem.pos[0], elem.pos[1] + str.bytesize] <=> pos) <= 0
-
-              str.each_line do |line|
-                if line.end_with?("\n")
-                  pos[0] += 1
-                  pos[1] = 0
-                else
-                  pos[1] += line.bytesize
-                end
-              end
-
-              yield(elem.event, str, elem.state)
+              on_scan.call(elem)
             end
           else
-            lexer.parse.each do |elem|
-              yield(elem.event, elem.tok, elem.state)
+            lexer.parse.sort_by(&:pos).each do |elem|
```

Comment:
Tests still pass without this change. Why is it needed?

Reply:
I added test case `<<A+1\nA` for this.

Another code `lexer.parse.reject` in ruby-lex.rb should also be fixed to `lexer.parse.sort_by(&:pos).reject`. This change is in another pull request #390

Reply:
I removed this `sort_by` again and the test still passed 🤔

Reply:
This `else` branch is only executed when ruby < 2.7 and test failed in my environment `ruby 2.6.0p0 (2018-12-25 revision 66547) [x86_64-darwin19]` without `sort_by`.

```
[tomoya:irb]% ruby -v
ruby 2.6.0p0 (2018-12-25 revision 66547) [x86_64-darwin19]
[tomoya:irb]% git log -n1
commit da54e7f0813d88546091af5bf220d83a6bd10198 (HEAD -> colorize_every_characters, tompng/colorize_every_characters)
Author: tompng <tomoyapenguin@gmail.com>
Date:   Mon Sep 19 14:14:10 2022 +0900

    Rewrite on_scan proc to be more readable.
[tomoya:irb]% git diff
[tomoya:irb]% rake test
...skip...
..........................
Finished in 6.179684 seconds.
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------
114 tests, 1158 assertions, 0 failures, 0 errors, 6 pendings, 0 omissions, 0 notifications
94.7368% passed
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------
18.45 tests/s, 187.39 assertions/s
[tomoya:irb]% # delete sort_by(&:pos)
[tomoya:irb]% git diff
diff --git a/lib/irb/color.rb b/lib/irb/color.rb
index 7071696..abe0590 100644
--- a/lib/irb/color.rb
+++ b/lib/irb/color.rb
@@ -196,7 +196,7 @@ module IRB # :nodoc:
               on_scan.call(elem)
             end
           else
-            lexer.parse.sort_by(&:pos).each do |elem|
+            lexer.parse.each do |elem|
               on_scan.call(elem)
             end
           end
[tomoya:irb]% rake test
...skip...
..................F
=========================================================================================================================================================================
Failure: test_colorize_code(TestIRB::TestColor)
/Users/tomoya/Documents/github/ruby/irb/test/irb/test_color.rb:269:in `assert_equal_with_term'
/Users/tomoya/Documents/github/ruby/irb/test/irb/test_color.rb:139:in `block in test_colorize_code'
     136: 
     137:       tests.each do |code, result|
     138:         if colorize_code_supported?
  => 139:           assert_equal_with_term(result, code, complete: true)
     140:           assert_equal_with_term(result, code, complete: false)
     141: 
     142:           assert_equal_with_term(code, code, complete: true, tty: false)
/Users/tomoya/Documents/github/ruby/irb/test/irb/test_color.rb:137:in `each'
/Users/tomoya/Documents/github/ruby/irb/test/irb/test_color.rb:137:in `test_colorize_code'
Case: colorize_code("<<A+1\nA", complete: true)
Result: "#{RED}<<A#{CLEAR}+1\n#{RED}A#{CLEAR}"
<"\e[31m<<A\e[0m+\e[34m\e[1m1\e[0m\n" + "\e[31mA\e[0m"> expected but was
<"\e[31m<<A\e[0m+1\n" + "\e[31mA\e[0m">

diff:
? <<A+1m1
  A
=========================================================================================================================================================================
...skip...
..........................
Finished in 4.4035 seconds.
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------
114 tests, 1103 assertions, 1 failures, 0 errors, 6 pendings, 0 omissions, 0 notifications
93.8596% passed
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------
25.89 tests/s, 250.48 assertions/s
rake aborted!
Command failed with status (1)
/Users/tomoya/.rbenv/versions/2.6.0/bin/bundle:23:in `load'
/Users/tomoya/.rbenv/versions/2.6.0/bin/bundle:23:in `<main>'
Tasks: TOP => test
(See full trace by running task with --trace)
```

Reply:
Ah sorry I missed that 🤦‍♂️